### DISCUSSION TITLE
`scan` over multiple modules?

### DISCUSSION THREAD

**Discussion by thisiscam** [google/flax#1340] [Q&A] [answered]
Hi,

I've been wondering if it's possible to leverage `scan` for the transformer stack:
https://github.com/google/flax/blob/master/examples/wmt/models.py#L353.

Based on my understanding, this should be possible with pure jax since all the shapes are fixed. 
Using scan should significantly speedup the compilation time for these kind of models.

Is there anyway to do this in flax? 

Here's what I have right now:
```python
Encoder1DBlockStack = nn.scan(
    Encoder1DBlock,
    variable_axes={
        "params": 0,
        "cache": 0
    },
    split_rngs={
        "params": True,
    },
    length=cfg.num_layers,
)
(x, _), _ = Encoder1DBlockStack(config=cfg, name="encoderblockstack")((x, encoder_mask), None)
```

Unfortunately, I believe I shouldn't pass `encoder_mask` in as a carry! 
In pure jax, I could basically capture `encoder_mask` from the outer scope of the scanned function, since it doesn't change.
Is there anyway achieve that? 

Thanks!

**Comment by jheek** [comment, marked as answer]
This should work. The first argument is indeed the carry. I think what you are missing is that you should use `in_axes=nn.broadcast` such that the encoder mask becomes broadcasted across the loop:


```
Encoder1DBlockStack = nn.scan(
    Encoder1DBlock,
    variable_axes={
        "params": 0,
        "cache": 0
    },
    split_rngs={
        "params": True,
    },
    length=cfg.num_layers,
    in_axes=(nn.broadcast,)
)
x, _ = Encoder1DBlockStack(config=cfg, name="encoderblockstack")(x, encoder_mask)
```

**Reply by thisiscam** [reply]
Thanks! This is what I ended up using:
```python
Encoder1DBlockStack = nn.scan(
      Encoder1DBlock,
      variable_axes={
          "params": 0,
          "cache": 0
      },
      split_rngs={
          "params": True,
          "dropout": True,
      },
      in_axes=nn.broadcast,
      length=cfg.num_layers,
  )
x, _ = Encoder1DBlockStack(config=cfg,
                                 name="encoderblockstack")(x, encoder_mask)
```

The compilation is much much faster now. 
If I understand correctly, this could have the potential downside of a slightly worse runtime performance? 
Maybe I can benchmark that, and if this is not worse at all, I can send PR to the examples to use this?

**Reply by jheek** [reply]
We have experimented with this internally especially for very deep networks but we aren't totally happy yet with the runtime overhead. For training especially you need to be careful and you might need to remat the scan body so `Encoder1DBlock` => `nn.remat(Encoder1DBlock)` to avoid OOM issues.

We will need better support for memory managment in scan and some more robust XLA support for optimizing loops to make this a free lunch.

**Reply by thisiscam** [reply]
Oh I see. That's unfortunate.
Do you have an idea of how long it takes to compile the wmt example?

**Reply by thisiscam** [reply]
> Do you have an idea of how long it takes to compile the wmt example?

Sorry, I didn't mean to make it read as a rhetorical question.

I wanted to ask: do you know how long it takes to compile (any) Transformer model with more than 6 layers? 
For example, the wmt example, or the [t5x example](https://github.com/google-research/google-research/tree/master/flax_models/t5x)?

I tried myself (with a slightly modified model, but to my understanding shouldn't change compile/runtime too much). It took something around 40 mins to compile the training step and predict step! I just wanted to check on the numbers just to understand if I messed up somewhere else.

**Reply by jheek** [reply]
>> 40 mins to compile the training step and predict step!

Wow that's a really long time. How deep is the model and what is the accelerator that you are compiling for?

**Reply by thisiscam** [reply]
It's a 6 layers transformer on TPU. 
I just tried on GPU and it is slightly faster, but still very slow compile (20 to 30 mins).
